### PR TITLE
CPD-6813: Nested error support.

### PR DESCRIPTION
**Pull request by AlkiviadisPapadakis-je**
This PR add supports to log error chains (created via the usage of the user info key `NSUnderlyingErrorKey`). Errors are now flattened out and put in the `errors` key of the log representation.

### CODE REVIEW

File: JustLog/Extensions/Dictionary+Flattening.swift [hunk 1]
```diff
@@ -7,47 +7,103 @@
...
+        }
         return retVal
     }
     
+   }
+
+extension Array where Element == Any {
+    
+    func flattened() -> [Any] {
+        var arraythings: [Any] = []
```
Comment (AlkiviadisPapadakis-je) [inline JustLog/Extensions/Dictionary+Flattening.swift]: I will rename this.

File: JustLog/Classes/Logger.swift [hunk 2]
```diff
@@ -43,6 +56,7 @@ public final class Logger: NSObject {
     public var enableFileLogging: Bool = true
     public var enableLogstashLogging: Bool = true
     public let internalLogger = SwiftyBeaver.self
+    public var eventLogingPolicy: ErrorLoggingPolicy = .singleEvent
```
Comment (ghost) [inline JustLog/Classes/Logger.swift]: typo in the word logging `eventLogingPolicy`. I propose `loggingPolicy`.

File: JustLog/Classes/Logger.swift [hunk 3]
```diff
@@ -105,47 +119,76 @@ extension Logger: Logging {
...
         let file = String(describing: file)
         let function = String(describing: function)
         let updatedUserInfo = [logTypeKey: "error"].merged(with: userInfo ?? [String : Any]())
-        let logMessage = self.logMessage(message, error: error, userInfo: updatedUserInfo, file, function, line)
-        internalLogger.error(logMessage, file, function, line: Int(line))
+        log(.error, message, error: error, userInfo: updatedUserInfo, file, function, line)
+    }
+    
+    internal func log(_ type: LogType, _ message: String, error: NSError?, userInfo: [String : Any]?, _ file: String, _ function: String, _ line: UInt) {
+        switch self.eventLogingPolicy {
```
Comment (ghost) [inline JustLog/Classes/Logger.swift]: The following lines

```
let logMessage = self.logMessage(message, error: error, userInfo: userInfo, file, function, line)
sendLogMessage(with: type, logMessage: logMessage, file, function, line)
```

are written 3 times. We could remove at least one occurrence like so:

```
switch loggingPolicy {
        case .multipleEvents where type == .error:
            if let error  = error {
                for error in error.disassociatedErrorChain() {
                    let logMessage = self.logMessage(message, error: error, userInfo: userInfo, file, function, line)
                    sendLogMessage(with: type, logMessage: logMessage, file, function, line)
                }
            }
        default:
            let logMessage = self.logMessage(message, error: error, userInfo: userInfo, file, function, line)
            sendLogMessage(with: type, logMessage: logMessage, file, function, line)
        }
```

File: JustLog/Classes/Logger.swift [hunk 4]
```diff
@@ -181,17 +224,25 @@ extension Logger {
         }
         
         if let error = error {
-            let errorInfo = [errorDomain: error.domain,
-                             errorCode: error.code] as [String : Any]
-            let errorUserInfo = error.humanReadableError().userInfo
-            options = options.merged(with: errorInfo).merged(with: errorUserInfo.flattened())
+            let keyMergePolicy: KeyMergePolicy = self.eventLogingPolicy == .singleEvent ? .encapsulateFlatten : .override
```
Comment (ghost) [inline JustLog/Classes/Logger.swift]: no need to use `self`

**Comment by AlkiviadisPapadakis-je** [comment]
@albertodebortoli-je I'm happy with the state of this. Ready to merge.

**Comment by AlkiviadisPapadakis-je** [comment]
I don't have access to remove the labels "Do not merge", "work in progress"

**Review by ghost — APPROVED** [review]

**Pull request merged**